renderdoc/driver/metal/metal_types.cpp

```diff
@@ -121,4 +123,499 @@ void DoSerialise(SerialiserType &ser, NS::String *&el)
+  SERIALISE_MEMBER(maxVertexAmplificationCount);
+  SERIALISE_MEMBER(colorAttachments);
+  SERIALISE_MEMBER(depthAttachmentPixelFormat);
+  SERIALISE_MEMBER(stencilAttachmentPixelFormat);
+  SERIALISE_MEMBER(inputPrimitiveTopology);
+  SERIALISE_MEMBER(tessellationPartitionMode);
+  SERIALISE_MEMBER(maxTessellationFactor);
+  SERIALISE_MEMBER(tessellationFactorScaleEnabled);
+  SERIALISE_MEMBER(tessellationFactorFormat);
+  SERIALISE_MEMBER(tessellationControlPointIndexType);
+  SERIALISE_MEMBER(tessellationFactorStepFunction);
+  SERIALISE_MEMBER(tessellationOutputWindingOrder);
+  SERIALISE_MEMBER(vertexBuffers);
+  SERIALISE_MEMBER(fragmentBuffers);
+  SERIALISE_MEMBER(supportIndirectCommandBuffers);
+  // TODO: will MTL::BinaryArchive need to be a wrapped resource
+  // SERIALISE_MEMBER(binaryArchives);
+  // TODO: will MTL::DynamicLibrary need to be a wrapped resource
+  // SERIALISE_MEMBER(vertexPreloadedLibraries);
+  // SERIALISE_MEMBER(fragmentPreloadedLibraries);
+  SERIALISE_MEMBER(vertexLinkedFunctions);
+  SERIALISE_MEMBER(fragmentLinkedFunctions);
+  SERIALISE_MEMBER(supportAddingVertexBinaryFunctions);
+  SERIALISE_MEMBER(supportAddingFragmentBinaryFunctions);
+  SERIALISE_MEMBER(maxVertexCallStackDepth);
+  SERIALISE_MEMBER(maxFragmentCallStackDepth);
+}
+
+static bool ValidData(MTL::VertexAttributeDescriptor *attribute)
+{
+  if(attribute->format() == MTL::VertexFormatInvalid)
+    return false;
+  return true;
+}
+
+static bool ValidData(MTL::VertexBufferLayoutDescriptor *layout)
+{
+  if(layout->stride() == 0)
+    return false;
+  return true;
+}
+
+static bool ValidData(MTL::PipelineBufferDescriptor *descriptor)
+{
+  if(descriptor->mutability() == MTL::MutabilityDefault)
+    return false;
+  return true;
+}
+
+static bool ValidData(MTL::RenderPipelineColorAttachmentDescriptor *descriptor)
+{
+  if(descriptor->pixelFormat() == MTL::PixelFormatInvalid)
+    return false;
+  return true;
+}
+
+namespace RDMTL
+{
+template <typename MTL_TYPE>
+static void GetWrappedNSArray(rdcarray<typename UnwrapHelper<MTL_TYPE>::Outer *> &to, NS::Array *from)
+{
+  int count = from->count();
+  to.resize(count);
+  for(int i = 0; i < count; ++i)
+  {
+    to[i] = GetWrapped((MTL_TYPE)from->object(i));
+  }
+}
+
+#define GETWRAPPEDNSARRAY(TYPE, NAME) GetWrappedNSArray<MTL::TYPE *>(NAME, objc->NAME())
+
+template <typename MTL_TYPE>
+static NS::Array *CreateUnwrappedNSArray(rdcarray<typename UnwrapHelper<MTL_TYPE>::Outer *> &from)
+{
+  int count = from.count();
+  if(count)
+  {
+    rdcarray<MTL_TYPE> unwrapped(count);
+    for(int i = 0; i < count; ++i)
+    {
+      unwrapped[i] = Unwrap(from[i]);
+    }
+    return NS::Array::array((NS::Object **)(unwrapped.data()), unwrapped.count());
+  }
+  return NULL;
+}
+
+template <typename RDMTL_TYPE, typename MTLARRAY_TYPE, typename MTL_TYPE, int MAX_COUNT>
+static void GetObjcArray(rdcarray<RDMTL_TYPE> &to, MTLARRAY_TYPE *from, bool (*validData)(MTL_TYPE *))
+{
+  MTL_TYPE *objcData[MAX_COUNT];
```

Comment:
What is the reason for this plain array here? Is there any reason you couldn't `push_back` to the `to` array in the loop below, up to `MAX_COUNT`, optionally with a `reserve` if you wanted? Or if metal's array has a way of querying the length, even better you could reserve/resize using that.